cmd/rook/ceph/sanitize_disk.go

```diff
@@ -0,0 +1,219 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ceph
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/daemon/ceph/osd"
+	oposd "github.com/rook/rook/pkg/operator/ceph/cluster/osd"
+)
+
+const (
+	ddBS    = "1M"           // DD's block size
+	ddCount = "10"           // DD runs over the first 10 offsets
+	ddFlags = "direct,dsync" // DD's sync flags"
+)
+
+// DiskSanitizer is simple struct to old the context to execute the commands
+type DiskSanitizer struct {
+	context          *clusterd.Context
+	sanitizeDiskSpec *cephv1.SanitizeDiskSpec
+	disk             string
+}
+
+func newDiskSanitizer(context *clusterd.Context) DiskSanitizer {
+	return DiskSanitizer{
+		context: context,
+		sanitizeDiskSpec: &cephv1.SanitizeDiskSpec{
+			SanitizeMethod:     cephv1.SanitizeMethodProperty(sanitizeMethod),
+			SanitizeUtility:    cephv1.SanitizeUtilityProperty(sanitizeUtility),
+			SanitizeDataSource: cephv1.SanitizeDataSourceProperty(sanitizeDataSource),
+			SanitizeIteration:  &sanitizeIteration,
+		},
+	}
+}
+
+func startSanitizeDisk(context *clusterd.Context) {
+	// Initialize sanitizer object
+	sanitizer := newDiskSanitizer(context)
+
+	// LVM based OSDs
+	osdLVMList, err := osd.GetCephVolumeLVMOSDs(context, clusterName, clusterFSID, "", false, false)
+	if err != nil {
+		logger.Errorf("failed to list lvm osd(s). %v", err)
+	} else {
+		// Start the sanitizing sequence
+		sanitizer.sanitizeLVMDisk(osdLVMList)
+	}
+
+	// Raw based OSDs
+	osdRawList, err := osd.GetCephVolumeRawOSDs(context, clusterName, clusterFSID, "", "", false)
+	if err != nil {
+		logger.Errorf("failed to list raw osd(s). %v", err)
+	} else {
+		// Start the sanitizing sequence
+		sanitizer.sanitizeRawDisk(osdRawList)
+	}
+}
+
+func (s *DiskSanitizer) sanitizeRawDisk(osdRawList []oposd.OSDInfo) {
+	// Initialize work group to wait for completion of all the go routine
+	var wg sync.WaitGroup
+
+	for _, osd := range osdRawList {
+		// Assign disk to sanitize
+		s.disk = osd.BlockPath
+
+		logger.Infof("sanitizing osd %d disk %q with %q", osd.ID, osd.BlockPath, s.sanitizeDiskSpec.SanitizeUtility.String())
+
+		// Increment the wait group counter
+		wg.Add(1)
+
+		// Put each sanitize in a go routine to speed things up
+		go s.executeCommand(&wg)
+	}
+
+	wg.Wait()
+}
+
+func (s *DiskSanitizer) sanitizeLVMDisk(osdLVMList []oposd.OSDInfo) {
+	// Initialize work group to wait for completion of all the go routine
+	var wg sync.WaitGroup
+	pvs := []string{}
+
+	for _, osd := range osdLVMList {
+		// Increment the wait group counter
+		wg.Add(1)
+
+		// Assign disk
+		s.disk = osd.BlockPath
```

Comment:
`disk` needs to be passed to the wipe method, otherwise you're going to overwrite this var that multiple goroutines are using.

Reply:
Do you mean as an arg? Because so far it has been working fine.